batch2/batch/driver/scheduler.py

```diff
@@ -21,6 +23,79 @@ def __init__(self, app):
         asyncio.ensure_future(self.loop('cancel_loop', self.cancel_state_changed, self.cancel_1))
         asyncio.ensure_future(self.bump_loop())
 
+    async def compute_fair_share(self):
+        free_cores_mcpu = sum([worker.free_cores_mcpu for worker in self.inst_pool.healthy_instances_by_free_cores])
+
+        user_running_cores_mcpu = {}
+        user_total_cores_mcpu = {}
+        user_allocated_cores = {}
+
+        pending_users_by_running_cores = sortedcontainers.SortedSet(
+            key=lambda user: user_running_cores_mcpu[user])
+        allocating_users_by_total_cores = sortedcontainers.SortedSet(
+            key=lambda user: user_total_cores_mcpu[user])
+
+        records = self.db.execute_and_fetchall(
+            '''
+SELECT user, ready_cores_mcpu, running_cores_mcpu
+FROM user_resources
+WHERE ready_cores_mcpu > 0;
+''')
+
+        async for record in records:
+            user = record['user']
+            user_running_cores_mcpu[user] = record['running_cores_mcpu']
+            user_total_cores_mcpu[user] = record['running_cores_mcpu'] + record['ready_cores_mcpu']
+            pending_users_by_running_cores.add(user)
+
+        def allocate_cores(user, mark):
+            user_allocated_cores[user] = int(mark - user_running_cores_mcpu[user] + 0.5)
+
+        mark = 0
+        while free_cores_mcpu > 0 and (pending_users_by_running_cores or allocating_users_by_total_cores):
+            lowest_running = None
+            lowest_total = None
+
+            if pending_users_by_running_cores:
+                lowest_running_user = pending_users_by_running_cores[0]
+                lowest_running = user_running_cores_mcpu[lowest_running_user]
+                if lowest_running == mark:
+                    pending_users_by_running_cores.remove(lowest_running_user)
+                    allocating_users_by_total_cores.add(lowest_running_user)
+                    continue
+
+            if allocating_users_by_total_cores:
+                lowest_total_user = allocating_users_by_total_cores[0]
+                lowest_total = user_total_cores_mcpu[lowest_total_user]
+                if lowest_total == mark:
+                    allocating_users_by_total_cores.remove(lowest_total_user)
+                    allocate_cores(lowest_total_user, mark)
+                    continue
+
+            if lowest_running is not None and lowest_total is not None:
+                allocation = min(lowest_running, lowest_total)
+            elif lowest_running is not None:
+                allocation = lowest_running
+            elif lowest_total is not None:
+                allocation = lowest_total
+            else:
+                break
+
+            n_allocating_users = len(allocating_users_by_total_cores)
+            cores_to_allocate = n_allocating_users * (allocation - mark)
+
+            if cores_to_allocate > free_cores_mcpu:
+                mark += free_cores_mcpu / n_allocating_users
```

Comment:
Ah, one more thing: we should make sure mark stays as an integer, since we use it in an equality comparison above.  I know this is on the exit, but makes me nervous anyway.  Just put the round here and remove it from allocate_cores.